File: app/src/main/java/com/github/benji377/timety/data/MainRepository.kt
package com.github.benji377.timety.data

import kotlinx.coroutines.flow.Flow

class MainRepository(private val database: AppDatabase) {
    // User
    val user: Flow<User?> = database.userDao().getUser()
    suspend fun insertOrUpdateUser(user: User) = database.userDao().insertOrUpdateUser(user)

    // Tasks
    val allTasks: Flow<List<Task>> = database.taskDao().getAllTasks()
    fun getTasksByStatus(status: TaskStatus): Flow<List<Task>> =
        database.taskDao().getTasksByStatus(status)

    fun getTasksInRange(start: Long, end: Long): Flow<List<Task>> =
        database.taskDao().getTasksInRange(start, end)

    suspend fun insertTask(task: Task) = database.taskDao().insertTask(task)
    suspend fun getTaskById(taskId: Int): Task? = database.taskDao().getTaskById(taskId)
    suspend fun updateTaskStatus(taskId: Int, status: TaskStatus) =
        database.taskDao().updateTaskStatus(taskId, status)

    suspend fun updateTask(task: Task) = database.taskDao().updateTask(task)
    suspend fun deleteTask(task: Task) = database.taskDao().deleteTask(task)
    suspend fun updateOverdueTasks(now: Long) =
        database.taskDao().updateOverdueTasks(now, TaskStatus.TODO, TaskStatus.OVERDUE)

    // Categories
    val allCategories: Flow<List<Category>> = database.categoryDao().getAllCategories()
    suspend fun insertCategory(category: Category) = database.categoryDao().insertCategory(category)
    suspend fun updateCategory(category: Category) = database.categoryDao().updateCategory(category)
    suspend fun deleteCategory(category: Category) = database.categoryDao().deleteCategory(category)

    // Focus Sessions
    val allSessions: Flow<List<FocusSession>> = database.focusSessionDao().getAllSessions()
    fun getSessionsForTask(taskId: Int) = database.focusSessionDao().getSessionsForTask(taskId)
    fun getSessionsByCategory(categoryId: Int) =
        database.focusSessionDao().getSessionsByCategory(categoryId)

    fun getSessionsForDay(startOfDay: Long, endOfDay: Long) =
        database.focusSessionDao().getSessionsForDay(startOfDay, endOfDay)

    suspend fun insertSession(session: FocusSession) =
        database.focusSessionDao().insertSession(session)

    suspend fun deleteSession(session: FocusSession) =
        database.focusSessionDao().deleteSession(session)

    // Focus Modes
    val allFocusModes: Flow<List<FocusMode>> = database.focusModeDao().getAllFocusModes()
    suspend fun insertFocusMode(focusMode: FocusMode) =
        database.focusModeDao().insertFocusMode(focusMode)

    suspend fun updateFocusMode(focusMode: FocusMode) =
        database.focusModeDao().updateFocusMode(focusMode)

    suspend fun deleteFocusMode(focusMode: FocusMode) =
        database.focusModeDao().deleteFocusMode(focusMode)

    // Daily Events
    fun getEventsForDay(startOfDay: Long, endOfDay: Long) =
        database.dailyEventDao().getEventsForDay(startOfDay, endOfDay)

    suspend fun insertEvent(event: DailyEvent) = database.dailyEventDao().insertEvent(event)
    suspend fun deleteEvent(event: DailyEvent) = database.dailyEventDao().deleteEvent(event)
}

File: lib/data/main_repository.dart
import 'package:sqflite/sqflite.dart';
import 'database_helper.dart';
import 'user.dart';
import 'task.dart';
import 'category.dart';
import 'focus_session.dart';
import 'focus_mode.dart';
import 'daily_event.dart';

class MainRepository {
  final DatabaseHelper _dbHelper = DatabaseHelper();

  // User
  Future<User?> getUser() async {
    final db = await _dbHelper.database;
    final List<Map<String, dynamic>> maps = await db.query('user', where: 'id = 1');
    if (maps.isEmpty) return null;
    return User.fromMap(maps.first);
  }

  Future<void> insertOrUpdateUser(User user) async {
    final db = await _dbHelper.database;
    await db.insert(
      'user',
      user.toMap(),
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  // Tasks
  Future<List<Task>> getAllTasks() async {
    final db = await _dbHelper.database;
    final List<Map<String, dynamic>> maps = await db.query('task');
    return maps.map((m) => Task.fromMap(m)).toList();
  }

  Future<List<Task>> getTasksByStatus(TaskStatus status) async {
    final db = await _dbHelper.database;
    final List<Map<String, dynamic>> maps = await db.query(
      'task',
      where: 'status = ?',
      whereArgs: [status.index],
    );
    return maps.map((m) => Task.fromMap(m)).toList();
  }

  Future<List<Task>> getTasksInRange(int startMillis, int endMillis) async {
    final db = await _dbHelper.database;
    final List<Map<String, dynamic>> maps = await db.query(
      'task',
      where: 'dueDate >= ? AND dueDate <= ?',
      whereArgs: [startMillis, endMillis],
    );
    return maps.map((m) => Task.fromMap(m)).toList();
  }

  Future<int> insertTask(Task task) async {
    final db = await _dbHelper.database;
    return await db.insert('task', task.toMap());
  }

  Future<Task?> getTaskById(int taskId) async {
    final db = await _dbHelper.database;
    final maps = await db.query('task', where: 'id = ?', whereArgs: [taskId]);
    if (maps.isEmpty) return null;
    return Task.fromMap(maps.first);
  }

  Future<void> updateTaskStatus(int taskId, TaskStatus status) async {
    final db = await _dbHelper.database;
    await db.update(
      'task',
      {'status': status.index},
      where: 'id = ?',
      whereArgs: [taskId],
    );
  }

  Future<void> updateTask(Task task) async {
    final db = await _dbHelper.database;
    await db.update(
      'task',
      task.toMap(),
      where: 'id = ?',
      whereArgs: [task.id],
    );
  }

  Future<void> deleteTask(int taskId) async {
    final db = await _dbHelper.database;
    await db.delete('task', where: 'id = ?', whereArgs: [taskId]);
  }

  Future<void> updateOverdueTasks(int nowMillis) async {
    final db = await _dbHelper.database;
    await db.update(
      'task',
      {'status': TaskStatus.overdue.index},
      where: 'dueDate < ? AND status = ?',
      whereArgs: [nowMillis, TaskStatus.todo.index],
    );
  }

  // Categories
  Future<List<Category>> getAllCategories() async {
    final db = await _dbHelper.database;
    final List<Map<String, dynamic>> maps = await db.query('category');
    return maps.map((m) => Category.fromMap(m)).toList();
  }

  Future<int> insertCategory(Category category) async {
    final db = await _dbHelper.database;
    return await db.insert('category', category.toMap());
  }

  Future<void> updateCategory(Category category) async {
    final db = await _dbHelper.database;
    await db.update(
      'category',
      category.toMap(),
      where: 'id = ?',
      whereArgs: [category.id],
    );
  }

  Future<void> deleteCategory(int categoryId) async {
    final db = await _dbHelper.database;
    await db.delete('category', where: 'id = ?', whereArgs: [categoryId]);
  }

  // Focus Sessions
  Future<List<FocusSession>> getAllSessions() async {
    final db = await _dbHelper.database;
    final List<Map<String, dynamic>> maps = await db.query('focus_session');
    return maps.map((m) => FocusSession.fromMap(m)).toList();
  }

  Future<List<FocusSession>> getSessionsForTask(int taskId) async {
    final db = await _dbHelper.database;
    final maps = await db.query('focus_session', where: 'taskId = ?', whereArgs: [taskId]);
    return maps.map((m) => FocusSession.fromMap(m)).toList();
  }

  Future<List<FocusSession>> getSessionsForDay(int startMillis, int endMillis) async {
    final db = await _dbHelper.database;
    final maps = await db.query(
      'focus_session',
      where: 'startTime >= ? AND startTime <= ?',
      whereArgs: [startMillis, endMillis],
    );
    return maps.map((m) => FocusSession.fromMap(m)).toList();
  }

  Future<int> insertSession(FocusSession session) async {
    final db = await _dbHelper.database;
    return await db.insert('focus_session', session.toMap());
  }

  // Focus Modes
  Future<List<FocusMode>> getAllFocusModes() async {
    final db = await _dbHelper.database;
    final maps = await db.query('focus_mode');
    return maps.map((m) => FocusMode.fromMap(m)).toList();
  }

  Future<int> insertFocusMode(FocusMode focusMode) async {
    final db = await _dbHelper.database;
    return await db.insert('focus_mode', focusMode.toMap());
  }

  Future<void> updateFocusMode(FocusMode focusMode) async {
    final db = await _dbHelper.database;
    await db.update(
      'focus_mode',
      focusMode.toMap(),
      where: 'id = ?',
      whereArgs: [focusMode.id],
    );
  }

  Future<void> deleteFocusMode(int id) async {
    final db = await _dbHelper.database;
    await db.delete('focus_mode', where: 'id = ?', whereArgs: [id]);
  }

  // Daily Events
  Future<List<DailyEvent>> getEventsForDay(int startMillis, int endMillis) async {
    final db = await _dbHelper.database;
    final maps = await db.query(
      'daily_event',
      where: 'timestamp >= ? AND timestamp <= ?',
      whereArgs: [startMillis, endMillis],
    );
    return maps.map((m) => DailyEvent.fromMap(m)).toList();
  }

  Future<int> insertEvent(DailyEvent event) async {
    final db = await _dbHelper.database;
    return await db.insert('daily_event', event.toMap());
  }
}
